go/consensus/cometbft/apps/roothash/state/state.go

```diff
@@ -270,6 +278,69 @@ func (s *ImmutableState) IncomingMessageQueue(ctx context.Context, runtimeID com
 	return msgs, nil
 }
 
+// PastRootsForRound returns the state and I/O roots for the given runtime ID and round.
+func (s *ImmutableState) PastRootsForRound(ctx context.Context, runtimeID common.Namespace, round uint64) (*roothash.PastRoots, error) {
+	raw, err := s.is.Get(ctx, pastRootsFmt.Encode(&runtimeID, round))
+	if err != nil {
+		return nil, api.UnavailableStateError(err)
+	}
+	if raw == nil {
+		// No roots present for given runtime and round.
+		return nil, nil
+	}
+
+	var roots roothash.PastRoots
+	if err = cbor.Unmarshal(raw, &roots); err != nil {
+		return nil, api.UnavailableStateError(err)
+	}
+	return &roots, nil
+}
+
+// PastRoots returns the past N state and I/O roots stored for the given
+// runtime ID.
+func (s *ImmutableState) PastRoots(ctx context.Context, runtimeID common.Namespace) (map[uint64]roothash.PastRoots, error) {
+	it := s.is.NewIterator(ctx)
+	defer it.Close()
+
+	// Round -> [state, I/O] roots.
+	ret := make(map[uint64]roothash.PastRoots)
+	for it.Seek(pastRootsFmt.Encode(&runtimeID)); it.Valid(); it.Next() {
+		var rtID keyformat.PreHashed
+		var round uint64
+		if !pastRootsFmt.Decode(it.Key(), &rtID, &round) {
+			break
+		}
+
+		var roots roothash.PastRoots
+		if err := cbor.Unmarshal(it.Value(), &roots); err != nil {
+			return nil, api.UnavailableStateError(err)
+		}
+
+		ret[round] = roots
+	}
+
+	return ret, nil
+}
+
+// PastRootsCount returns the number of past state and I/O roots in storage
+// for the given runtime ID.
```

Comment:
Maybe mention that this is more efficient than calling `len(s.PastRoots(id))` as it avoids deserialization.